### ISSUE TITLE
Backlog For UpdateEngine

### ISSUE THREAD

**Issue by ryanelian** [Arks-Layer/PSO2Tweaker#37]
Just noting these down so I don't forget:
- Prepatch implementation phase 1: Allow sideloading patch files downloaded into an offline repository before online update happens.
- Prepatch implementation phase 2: Implementation for actually downloading prepatch files from SEGA into the offline repository.
- Unrelated to `UpdateEngine` but noting here down anyway: Develop `PatchEngine` for dealing with patches. English patch looks easy enough to do first (simply unarchiving zip files). Integrate 7zip and Unrar executable functions into the engine using managed libraries.
- Revert Terminal to .NET Core project instead of the classic C# project. This allows direct linking against the core project references for easier debugging (instead of a separate Scratchpad project. Revert `PSO2 Tweaker.sln` to include only Tweaker.
- Formulate unit tests for the other modules.
-  Unlike `OnUpdateStart`, `OnUpdateCompleted` may not be called if there the method doesn't make it all the way to the end, due to exceptions or whatever. Consider renaming this into `OnUpdateSuccess` and then renaming `OnPatchingFailed` to `OnUpdateFailed`, baka ryan.
- Deprecate the `client.json` into something else: `game.json` or `pso2.json` to prevent shitstorm during future deployment due to people stopping the update mid-way.
- Add censor file to the blacklist. :cake: 
## Dones
- Game version updating. Currently the Tweaker updates the game version upon update completion. This feature should be done by the `UpdateEngine` instead via relevant `RegistryTweakerSettings` property implementation.
- `Abstractions`: Further granularize the `IRenderer` methods. `Writeline` method will be removed. This is a technical debt to be done so that the Tweaker code base is clean from all the string filtering filth and so that Excellion has more fine-grained control over the events that happened during the update.
- Improve UI responsiveness by offloading backup restore to background thread then awaiting it.

**Comment by TheCrimsonDevil** [comment]
>the string filtering filth 

feelsbadman

>Excellion

wut

**Comment by ryanelian** [comment]
> > Excellion

> wut

Contrary to beliefs, `UpdateEngine` and `Abstractions` were not taken apart from Tweaker but from an almost-abandoned pet project of mine called Excellion.

Excellion was developed by decompiling the Tweaker before it went open source then reverse engineering it, and before I went haitus from the game for months. The goal of Excellion is to be a fast and lightweight, personal-use, 20KB WPF application executable drop-in replacement for the Tweaker.

So far the only modules developed for the Excellion are: `UpdateEngine`, `Abstractions` and `CloudLogging`.

I don't intend to release nor support Excellion to public but it's being actively developed ever since I returned into the game.
